app/src/ui/merge-conflicts/merge-conflicts-dialog.tsx

```diff
@@ -0,0 +1,227 @@
+
+interface IMergeConflictsDialogProps {
+  readonly dispatcher: Dispatcher
+  readonly repository: Repository
+  readonly status: WorkingDirectoryStatus
+  readonly onDismissed: () => void
+  readonly openFileInExternalEditor: (path: string) => void
+  readonly externalEditorName?: string
+  readonly openRepositoryInShell: (repository: Repository) => void
+  readonly currentBranch: Branch
+  readonly comparisonBranch: Branch
+}
+
+const submitButtonString = 'Commit merge'
+const cancelButtonString = 'Abort merge'
+
+/**
+ * Modal to tell the user their merge encountered conflicts
+ */
+export class MergeConflictsDialog extends React.Component<
+  IMergeConflictsDialogProps,
+  {}
+> {
+  /**
+   *  commits the merge displays the repository changes tab and dismisses the modal
+   */
+  private onSubmit = async () => {
+    await createMergeCommit(this.props.repository, this.props.status.files)
+    this.props.dispatcher.changeRepositorySection(
+      this.props.repository,
+      RepositorySectionTab.Changes
+    )
+    this.props.onDismissed()
+  }
+
+  /**
+   *  dismisses the modal and shows the abort merge warning modal
+   */
+  private onCancel = async () => {
+    const anyResolvedFiles = this.getUnmergedFiles().some(
+      f => f.status === AppFileStatus.Resolved
+    )
+    if (!anyResolvedFiles) {
+      await abortMerge(this.props.repository)
+      this.props.onDismissed()
+    } else {
+      this.props.onDismissed()
+      this.props.dispatcher.showPopup({
+        type: PopupType.AbortMerge,
+        repository: this.props.repository,
+      })
+    }
+  }
+
+  private titleString(currentBranchName: string, comparisonBranchName: string) {
+    return `Resolve conflicts before merging ${comparisonBranchName} into ${currentBranchName}`
+  }
+
+  private editorButtonString(editorName: string | undefined) {
+    const defaultEditorString = 'editor'
+    return `Open in ${editorName || defaultEditorString}`
+  }
+
+  private renderShellLink(openThisRepositoryInShell: () => void): JSX.Element {
+    return (
+      <div className="cli-link">
+        You can also{' '}
+        <a onClick={openThisRepositoryInShell}>open the command line</a> to
+        resolve
+      </div>
+    )
+  }
+
+  private renderResolvedFile(path: string): JSX.Element {
+    return (
+      <li className="unmerged-file-status-resolved">
+        <Octicon symbol={OcticonSymbol.fileCode} className="file-octicon" />
+        <div className="column-left">
+          <PathText path={path} availableWidth={50} />
+          <div className="file-conflicts-status">No conflicts remaining</div>
+        </div>
+        <div className="green-circle">
+          <Octicon symbol={OcticonSymbol.check} />
+        </div>
+      </li>
+    )
+  }
+
+  private renderConflictedFile(
+    path: string,
+    conflicts: number,
+    editorName: string | undefined,
+    onOpenEditorClick: () => void
+  ): JSX.Element {
+    const humanReadableConflicts = Math.ceil(conflicts / 3)
```

Comment:
The `3` here is just a magic number that corresponds to diff start/middle/end markers, right?

 - `<<<<<<`
 - `======`
 - `>>>>>>`

Can we add a comment here to explain why we're computing the number in this way?